server/src/main/java/io/horizen/services/impl/NscServiceImpl.java

```diff
@@ -0,0 +1,192 @@
+package io.horizen.services.impl;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import com.google.inject.Inject;
+import io.horizen.config.Settings;
+import io.horizen.helpers.Helper;
+import io.horizen.helpers.MyGsonManager;
+import io.horizen.services.NscService;
+import org.web3j.abi.TypeDecoder;
+import org.web3j.abi.datatypes.Address;
+import org.web3j.abi.datatypes.generated.Bytes3;
+import org.web3j.abi.datatypes.generated.Bytes32;
+import org.web3j.abi.datatypes.generated.Uint32;
+import org.web3j.crypto.Hash;
+import org.web3j.crypto.Keys;
+import org.web3j.utils.Numeric;
+import spark.utils.IOUtils;
+
+import java.net.HttpURLConnection;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.*;
+import java.util.regex.Pattern;
+
+public class NscServiceImpl implements NscService {
+    private final Settings settings;
+
+    @Inject
+    public NscServiceImpl(Settings settings) {
+        this.settings = settings;
+    }
+
+    public Map<String, List<String>> getNscOwnerships(String scAddress) throws Exception {
+        String method;
+        String abiString;
+        if (scAddress == null) {
+            method = "getAllKeyOwnerships()";
+            // Generate the 4-byte selector
+            byte[] selector = Arrays.copyOf(Hash.sha3(method.getBytes()),4);
+            abiString = "0x" + Numeric.toHexStringNoPrefix(selector);
+        } else {
+            // Remove the "0x" prefix
+            if (scAddress.startsWith("0x"))
+                scAddress = scAddress.substring(2);
+
+            if (scAddress.length() != 40 || !Pattern.matches("[0-9A-Fa-f]+", scAddress))
+                throw new Exception("Invalid sc address length: {}, expected 40");
+
+            method = "getKeyOwnerships(address)";
+            byte[] selector = Arrays.copyOf(Hash.sha3(method.getBytes()),4);
+            String abiMethodString = "0x" + Numeric.toHexStringNoPrefix(selector);
+            abiString = abiMethodString + "000000000000000000000000" + scAddress;
+        }
+
+        String requestBody = buildNscRequestBody(abiString);
+        HttpURLConnection connection = Helper.sendRequestWithAuth(settings.getNscUrl() + "ethv1", requestBody, "user", "Horizen");
```

Comment:
We should probably set some constant definition in the settings for these 3 parameters

Reply:
Done